File: Solution/11.container-with-most-water.go
package LeetCode

/*
 * @lc app=leetcode id=11 lang=golang
 *
 * [11] Container With Most Water
 */

// @lc code=start
func maxArea(height []int) int {
	if len(height) <= 1 {
		return 0
	}

	// Initialize variables for storing max area and the two pointers
	var max int
	// var max float64

	// Iterate through the array while the two pointers don't overlap
	for left, right := 0, len(height)-1; left <= right; {
		// Calculate the area between the two pointers and update max if necessary
		if height[left] < height[right] {
			max = Max(max, height[left]*(right-left))
			// max = math.Max(max, float64(height[left]*(right-left)))
			left++
		} else {
			max = Max(max, height[right]*(right-left))
			// max = math.Max(max, float64(height[right]*(right-left)))
			right--
		}
	}

	// Return the maximum area
	return max
}

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

// @lc code=end

File: 11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python
#
# [11] Container With Most Water
#


# @lc code=start
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if len(height) <= 1:
            return 0

        max_area = 0

        left, right = 0, len(height) - 1
        while left < right:
            if height[left] < height[right]:
                max_area = max(max_area, height[left] * (right - left))
                left += 1
            else:
                max_area = max(max_area, height[right] * (right - left))
                right -= 1

        return max_area


# @lc code=end
